### ISSUE TITLE
Remote node IP addresses kept in ipcaches when the node is removed

### ISSUE THREAD

**Issue by pchaigno** [cilium/cilium#21844]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

When the cluster is resized and a node is removed, its IP addresses are kept in the ipcache maps of other nodes. Those IP addresses are then overwritten when they are reallocated.

```
$ gcloud container clusters create paul-cluster --num-nodes 3 --machine-type n1-standard-4 --zone europe-west4-a --node-taints node.cilium.io/agent-not-ready=true:NoSchedule
$ cilium install
# Assuming cilium-lkbnc is on the removed node:
$ kubectl -n kube-system -c cilium-agent exec cilium-lkbnc -- ip a show dev cilium_host
    inet 10.68.0.82/32 scope link cilium_host
$ kubectl -n kube-system -c cilium-agent exec cilium-lkbnc -- ip a show dev eth0
    inet 10.164.0.29/32 scope global dynamic eth0
$ gcloud container clusters resize paul-cluster --num-nodes 2 --region europe-west4-a
# Assuming cilium-8btml is on one of the remaining nodes:
$ kubectl -n kube-system -c cilium-agent exec cilium-8btml -- cilium bpf ipcache get 10.68.0.82
10.68.0.82 maps to identity identity=6 encryptkey=3 tunnelendpoint=10.164.0.29
$ kubectl -n kube-system -c cilium-agent exec cilium-8btml -- cilium bpf ipcache get 10.164.0.29
10.164.0.29 maps to identity identity=6 encryptkey=0 tunnelendpoint=10.164.0.29
```

### Cilium Version

latest (4bafe1750f)

### Kernel Version

5.10.133 (Google COS)

### Kubernetes Version

v1.22.12-gke.2300

### Sysdump

[cilium-sysdump-20221021-165641.zip](https://github.com/cilium/cilium/files/9840184/cilium-sysdump-20221021-165641.zip)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aditighag** [comment]
This behavior is mentioned in the overall improvement issue - https://github.com/cilium/cilium/issues/21725. Although the issue doesn't track individual work items, there can be other consumers of node events. We should evaluate if it's feasible to have a common place to identify stale nodes, and then propagate that information to the consumers.

**Comment by pchaigno** [comment]
Note in the present case, we receive a k8s event for the deleted node, so we should be able to clean the state based on that very easily.

**Comment by aditighag** [comment]
> Note in the present case, we receive a k8s event for the deleted node, so we should be able to clean the state based on that very easily.

From a cursory glance, that should already happen - https://github.com/cilium/cilium/blob/master/pkg/node/manager/manager.go#L633-L633. There are couple of checks that can skip the ipcache deletion though, but is that the case here?

**Comment by pchaigno** [comment]
I confirmed that commit 454df0c065bd94aa30a9d703732d64a24d3b6073 introduced this bug. I'll send a fix.

**Comment by carloscastrojumo** [comment]
@pchaigno Is this issue only present in `1.13`? We are seeing a similar behaviour in `1.12.4` but supposedly this version does not contains the commit [https://github.com/cilium/cilium/commit/454df0c065bd94aa30a9d703732d64a24d3b6073](https://github.com/cilium/cilium/commit/454df0c065bd94aa30a9d703732d64a24d3b6073) that introduced the issue.

**Comment by pchaigno** [comment]
@carloscastrojumo AFAIK, yes, it shouldn't be present in v1.12. And it should be fix for v1.13 (not yet released). If you're encountering a bug, please open a new issue.
